Sources/PointFree/Auth.swift

```diff
@@ -0,0 +1,123 @@
+import Either
+import Foundation
+import Html
+import HttpPipeline
+import HttpPipelineHtmlSupport
+import Optics
+import Prelude
+
+let secretHomeResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data?>> =
+  writeStatus(.ok)
+    >>> readGitHubSessionCookieMiddleware
+    >>> respond(secretHomeView)
+    >>> pure
+
+let githubCallbackResponse =
+  authTokenMiddleware
+    >-> redirect(to: link(to: .secretHome), headersMiddleware: writeGitHubSessionCookieMiddleware)
+    >>> pure
+
+let loginResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data?>> =
+  redirect(to: githubAuthorizationUrl)
+    >>> pure
+
+let logoutResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data?>> =
+  redirect(
+    to: link(to: .secretHome),
+    headersMiddleware: writeHeader(.clearCookie(key: "github_session"))
+    )
+    >>> pure
+
+private let secretHomeView = View<Either<Prelude.Unit, GitHubUserEnvelope>> { data in
+  [
+    p(["welcome home"]),
+
+    p([
+      text(
+        data.right.map { "You are logged in as \($0.gitHubUser.name)" }
+          ?? "You are not logged in"
+      )
+      ]),
+
+    p([
+       data.isRight
+        ? a([href(link(to: .logout))], ["Log out"])
+        : a([href(link(to: .login))], ["Log in"])
+      ])
+    ]
+}
+
+// todo: move to swift-web
+extension URLRequest {
+  public var cookies: [String: String] {
+    let pairs = (self.allHTTPHeaderFields?["Cookie"] ?? "")
+      .components(separatedBy: "; ")
+      .map {
+        $0.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
+          .map(String.init)
+      }
+      .flatMap { pure(createTuple) <*> $0.first <*> $0.last }
+    return .init(uniqueKeysWithValues: pairs)
+  }
+}
+
+// todo: move to prelude
+private func createTuple<A, B>(_ a: A) -> (B) -> (A, B) {
```

Comment:
should prob move this to the prelude, can be handy to do things like `pure(createTuple) <*> a <*> b`

Reply:
Haha, yep. I've been inlining `curry { ($0, $1) } <¢> a <*> b`.

Reply:
oh yeah, also turns out you have a `tuple: (A) -> (B) -> (A, B)` in the prelude already! we should make more, and we need some for `Tuple`.